readseeker.go

```diff
@@ -0,0 +1,153 @@
+
+	curChunkID     ChunkID // hash of current chunk
+	curChunk       []byte  // decompressed version of current chunk
+	curChunkIdx    int     // identity of current chunk
+	curChunkOffset int64   // offset within current chunk
+}
+
+func NewIndexReadSeeker(i Index, s Store) IndexPos {
+	lastChunk := i.Chunks[len(i.Chunks)-1]
+	length := int64(lastChunk.Start + lastChunk.Size)
+
+	return IndexPos{s, i, length, 0, i.Chunks[0].ID, nil, 0, 0}
+}
+
+/* findOffset - Actually update our IndexPos for a new Index
+ *
+ * - Seek forward within existing chunk if appropriate
+ * - Bisect the Chunks array to find the correct chunk
+ * - Decompress if id does not match curChunk
+ * - Update chunkIdx and chunkOffset
+ */
+func (ip *IndexPos) findOffset(newPos int64) (int64, error) {
+	var newChunkIdx int
+	var newChunkOffset int64
+	var delta int64
+	var err error = nil
+
+	// Degenerate case: Seeking to existing position
+	delta = newPos - ip.pos
+	if delta == 0 {
+		return ip.pos, nil
+	}
+
+	// Degenerate case: Seeking within current chunk
+	if (delta + ip.curChunkOffset) < int64(ip.Index.Chunks[ip.curChunkIdx].Size) {
+		ip.pos += delta
+		ip.curChunkOffset += delta
+		return ip.pos, nil
+	}
+
+	// General case: Bisect
+	chunks := ip.Index.Chunks
+	newChunkIdx = sort.Search(len(chunks), func(i int) bool { return int64(chunks[i].Start) >= newPos })
+	if newChunkIdx >= len(chunks) { // function was not true for any chunk -- meaning we're running off the end
+		newChunkIdx = len(chunks) - 1
+	}
+	newChunk := ip.Index.Chunks[newChunkIdx]
+	newChunkOffset = newPos - int64(newChunk.Start)
+
+	if newPos < int64(newChunk.Start) {
+		return ip.pos, fmt.Errorf("Seek found chunk beginning at position %v, desired position is %v", newChunk.Start, newPos)
+	}
+	if newPos > int64(newChunk.Start+newChunk.Size) {
+		return ip.pos, fmt.Errorf("Seek found chunk ending at position %v, desired position is %v", newChunk.Start+newChunk.Size, newPos)
+	}
+
+	// Only invalidate cache if new chunk is different from old one (avoid re-decompressing all-0 regions)
+	if newChunk.ID != ip.curChunkID {
+		ip.curChunk = nil // next read attempt will call loadChunk()
+	}
+	// BELOW HERE, WE HAVE UPDATED THE DATA AND MUST NOT ERROR
+	ip.curChunkIdx = newChunkIdx
+	ip.curChunkID = newChunk.ID
+	ip.curChunkOffset = newChunkOffset
+	ip.pos = newPos
+	return newPos, err
+}
+
+func (ip *IndexPos) loadChunk() (err error) {
+	var compressedChunk []byte
+	var decompressedChunk []byte
+
+	compressedChunk, err = ip.Store.GetChunk(ip.curChunkID)
+	if err != nil {
+		return err
+	}
+
+	decompressedChunk, err = Decompress(nil, compressedChunk)
+	if err != nil {
+		return err
+	}
+
+	ip.curChunk = decompressedChunk
+	return nil
+}
+
+func (ip *IndexPos) Seek(offset int64, whence int) (int64, error) {
+	var newPos int64
+	var err error
+	switch whence {
+	case io.SeekStart:
+		newPos = offset
+	case io.SeekCurrent:
+		newPos = ip.pos + offset
+	case io.SeekEnd:
+		newPos = ip.Length - offset
```

Comment:
I believe this should be:
```go
newPos = ip.Length + offset
```
but since that doesn't really work for this use case, perhaps simply return EOF

Reply:
This is a place where I misread the interface we were implementing early on, and didn't catch a consequence of that when fixing it. Thought offset was uint64, and thus that adding would make no sense; since it's int64, we can move backwards, so `+` is indeed correct.